### PR TITLE
gvfs: support cross compilation

### PR DESCRIPTION
**Pull request by uninsane**
## Description of changes

note that the gvfs [patch](https://gitlab.gnome.org/GNOME/gvfs/-/commit/8327383e262e1e7f32750a8a2d3dd708195b0f53.patch) nixpkgs presently reverts is incompatible with cross compilation because it hardcodes the *build* platform's `ssh` into the binary. the new `postPatch` code achieves the same function as that reverted patch but in a way that lets us use the host platform's ssh.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/gvfs/default.nix [hunk 1]
```diff
@@ -52,21 +52,16 @@ stdenv.mkDerivation rec {
...
   postPatch = ''
     # patchShebangs requires executable file
     chmod +x meson_post_install.py
     patchShebangs meson_post_install.py
     patchShebangs test test-driver
+
+    # Hardcode the ssh path again, but in a way which works for cross compilation.
+    # https://gitlab.gnome.org/GNOME/gvfs/-/issues/465
+    substituteInPlace daemon/meson.build \
+      --replace '-DSSH_PROGRAM="ssh"' '-DSSH_PROGRAM="${lib.getBin openssh}/bin/ssh"'
```
Comment (bjornfor) [inline pkgs/development/libraries/gvfs/default.nix:64]: What about the `SFTP backend` part of the patch this replaces? Doesn't that need to be handled anymore?

Comment (uninsane) [reply pkgs/development/libraries/gvfs/default.nix:64]: shouldn't be necessary:
```
if enable_sftp
  ssh = find_program('ssh', required: false)
  assert(ssh.found(), 'SFTP backend requested but a ssh client is required')
endif
```

this part of the patch was just a safeguard/diagnostic AFAICT. nothing else in any of the build files references that `ssh` variable (nor does any file in the repo reference the bare `"ssh"`, everything does properly go through `SSH_PROGRAM`)

**Review by bjornfor — APPROVED** [review]
I prefer having a patch that introduces meta-variables + search-replace, because only search-replace can easily bitrot. But it's not a blocker.

**Comment by uninsane** [comment]
@bjornfor i could accommodate that. easiest way would be to keep the same patch that we fetch now, and then in postPatch substituteInPlace the host ssh over this line `ssh = find_program('ssh', required: false)`.

it’s a weird patching-a-patch type of thing though: i could instead fold those together and store the single .patch file in nixpkgs. thoughts?

**Comment by bjornfor** [comment]
> substituteInPlace the host ssh over this line ssh = find_program('ssh', required: false)

That seems brittle. I'd rather have the patch create `@PATTERN_FOR_SEARCH_REPLACE@` and then run substituteInPlace on that.

> it’s a weird patching-a-patch type of thing though: i could instead fold those together and store the single .patch file in nixpkgs. thoughts?

Perhaps it's weird, but it's robust.

Committing a patch to nixpkgs can also work. A variant of this, if the patch is small enough, it can probably just as well be inlined in the expression:

```
let
  sshPatch = builtins.toFile "ssh.patch" ''
    - foo
    + @foo@
  '';
in ...
```

The benefit here is that it's not so easy to forget removing it if it becomes unneeded in the future.

**Comment by uninsane** [comment]
pushed a new diff which hardcodes the ssh path in a way that should be robust

**Review by bjornfor — APPROVED** [review]

**Pull request merged**